Enable IMDSv2 on AWSIAMCredentials

db725-bb opened the pull request:
Enables AWSIAMCredentials to work with IMDSv2

More info

https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html

db725-bb committed:
Enable IMDSv2 on AWSIAMCredentials

db725-bb committed:
PR feedback - change cache update

Adds a timestamp field to renew the cache based on expiration time instead of waiting for an unauthorized response

db725-bb committed:
PR feedback

Add IMDSv2Timestamp to shared cache, minor format fixes

db725-bb committed:
Format fixes

Remove spaces

paulpearcy commented on src/lua/api-gateway/aws/AWSIAMCredentials.lua:
```diff
@@ -189,4 +208,31 @@ function AWSIAMCredentials:getSecurityCredentials()
+-- Retrieves a IMDSv2 Token if it hasn't been done yet or if the previous token has expired
+function AWSIAMCredentials:retrieveIMDSv2Token()
+
+    if (cache.ExpireIMDSv2AtTimestamp == nil or cache.ExpireIMDSv2AtTimestamp - os.time() <= 0 ) then
```
Should this have a little fudge factor so the token doesn't expire right after this check? Maybe if it expires in 10 seconds?

paulpearcy commented on src/lua/api-gateway/aws/AWSIAMCredentials.lua:
```diff
@@ -103,13 +109,20 @@ end
     ngx.log(ngx.DEBUG, "Fetching IAM User from:",
         self.security_credentials_host, ":", self.security_credentials_port, self.security_credentials_url)
+    
+-- Retrieves the token needed for request under IMDSV2 
```
Minor nit, some of the indentation is off here and below.

db725-bb committed:
Add fudge factor in token expiration

10 seconds of margin to avoid edge case of token immediately expiring after check.
